src/pkg/p2p/preheat/job.go

```diff
@@ -0,0 +1,225 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package preheat
+
+import (
+	"time"
+
+	"github.com/goharbor/harbor/src/jobservice/job"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/pkg/p2p/preheat/models/provider"
+	pr "github.com/goharbor/harbor/src/pkg/p2p/preheat/provider"
+	"github.com/goharbor/harbor/src/pkg/p2p/preheat/provider/auth"
+)
+
+const (
+	// parameter keeps the preheating provider instance info.
+	preheatParamProvider = "provider"
+	// parameter keeps the preheating artifact (image) info.
+	preheatParamImage = "image"
+	// checkInterval indicates the interval of loop check.
+	checkInterval = 10 * time.Second
+	// checkTimeout indicates the overall timeout of the loop check.
+	checkTimeout = 1801 * time.Second
+)
+
+// Job preheats the given artifact(image) to the target preheat provider.
+type Job struct{}
+
+// MaxFails of preheat job. Don't need to retry.
+func (j *Job) MaxFails() uint {
+	return 1
+}
+
+// MaxCurrency indicates no limitation to the concurrency of preheat job.
+func (j *Job) MaxCurrency() uint {
+	return 0
+}
+
+// ShouldRetry indicates no need to retry preheat job as it's just for a cache purpose.
+func (j *Job) ShouldRetry() bool {
+	return false
+}
+
+// Validate the parameters of preheat job.
+func (j *Job) Validate(params job.Parameters) error {
+	_, err := parseParamProvider(params)
+	if err != nil {
+		return err
+	}
+
+	_, err = parseParamImage(params)
+
+	return err
+}
+
+// Run the preheat process.
+func (j *Job) Run(ctx job.Context, params job.Parameters) error {
+	// Get logger
+	myLogger := ctx.GetLogger()
+
+	// Parse parameters, ignore errors as they have been validated already
+	p, _ := parseParamProvider(params)
+	pi, _ := parseParamImage(params)
+
+	// Print related info to log first
+	myLogger.Infof(
+		"Preheating image '%s:%s' to the target preheat provider: %s %s:%s",
+		pi.ImageName,
+		pi.Tag,
+		p.Vendor,
+		p.Name,
+		p.Endpoint,
+	)
+
+	// Get driver factory for the given provider
+	fac, ok := pr.GetProvider(p.Vendor)
+	if !ok {
+		err := errors.Errorf("No driver registered for provider %s", p.Vendor)
+		myLogger.Error(err)
+		return preheatJobRunningError(err)
+	}
+
+	// Construct driver
+	d, err := fac(p)
+	if err != nil {
+		myLogger.Error(err)
+		return preheatJobRunningError(err)
+	}
+
+	myLogger.Infof("Get preheat provider driver: %s", p.Vendor)
+
+	// Start the preheat process
+	// First, check the health of the provider
+	h, err := d.GetHealth()
+	if err != nil {
+		return preheatJobRunningError(err)
```

Comment:
maybe we should record all error info in job log. so users can check the failure reason from UI.

Reply:
correct, missing the logs here